File: frontend/src/types/stock.ts
/** TypeScript types matching the backend Pydantic models. */

export interface SearchResult {
  symbol: string;
  name: string;
  exchange: string | null;
  type: string | null;
}

export interface SearchResponse {
  results: SearchResult[];
}

export interface CompanyProfile {
  symbol: string;
  name: string;
  sector: string | null;
  industry: string | null;
  country: string | null;
  website: string | null;
  description: string | null;
  market_cap: number | null;
  employees: number | null;
  currency: string | null;
  exchange: string | null;
  logo_url: string | null;
  shares_outstanding: number | null;
  current_price: number | null;
}

export interface PricePoint {
  date: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number | null;
  volume: number | null;
}

export interface HistoricalPrices {
  symbol: string;
  period: string;
  prices: PricePoint[];
}

export interface FinancialStatement {
  date: string;
  data: Record<string, number | null>;
}

export interface FinancialStatementResponse {
  symbol: string;
  statement_type: string;
  statements: FinancialStatement[];
}

export interface KeyRatios {
  symbol: string;
  pe_ratio: number | null;
  forward_pe: number | null;
  peg_ratio: number | null;
  price_to_book: number | null;
  price_to_sales: number | null;
  ev_to_ebitda: number | null;
  profit_margin: number | null;
  operating_margin: number | null;
  return_on_equity: number | null;
  return_on_assets: number | null;
  debt_to_equity: number | null;
  current_ratio: number | null;
  quick_ratio: number | null;
  dividend_yield: number | null;
  beta: number | null;
  fifty_two_week_high: number | null;
  fifty_two_week_low: number | null;
}

/** A saved projection/scenario from Supabase. */
export interface Projection {
  id: string;
  ticker: string;
  title: string;
  data: ProjectionData;
  created_at: string;
  updated_at: string;
}

/** The JSONB payload stored in a projection's data field. */
export type ProjectionData =
  | { model_type: "dcf"; inputs: DcfInputs; results: DcfResults }
  | { model_type: "multiples"; inputs: MultiplesInputs; results: MultiplesResults };

/** DCF model user-adjustable inputs — all assumption fields are per-year arrays. */
export interface DcfInputs {
  revenue_growth_rates: number[];
  operating_margins: number[];
  tax_rates: number[];
  capex_pct_revenues: number[];
  wacc: number;
  terminal_growth_rate: number;
  shares_outstanding: number;
  base_revenue: number;
}

/** Detailed computed values for a single projected year. */
export interface DcfYearDetail {
  revenue: number;
  revenue_growth: number;
  operating_income: number;
  operating_margin: number;
  taxes: number;
  tax_rate: number;
  nopat: number;
  capex: number;
  capex_pct: number;
  fcf: number;
  discount_factor: number;
  pv_fcf: number;
}

/** DCF model computed outputs. */
export interface DcfResults {
  years: DcfYearDetail[];
  projected_revenues: number[];
  projected_fcfs: number[];
  terminal_value: number;
  pv_terminal_value: number;
  enterprise_value: number;
  implied_share_price: number;
  current_price: number;
  upside_pct: number;
}

/** Averaged financial metrics across industry peers (from backend). */
export interface IndustryAverages {
  industry: string;
  peer_count: number;
  operating_margin: number | null;
  tax_rate: number | null;
  capex_pct_revenue: number | null;
  revenue_growth: number | null;
}

/** Company's own historical averages computed from financial statements. */
export interface CompanyAverages {
  operating_margin: number | null;
  tax_rate: number | null;
  capex_pct_revenue: number | null;
  revenue_growth: number | null;
}

/** Multiples model user-adjustable inputs. */
export interface MultiplesInputs {
  multiple_type: "pe" | "ev_revenue" | "ev_ebitda";
  projected_metric_value: number;
  target_multiple: number;
  shares_outstanding: number;
  net_debt: number;
}

/** Multiples model computed outputs. */
export interface MultiplesResults {
  implied_value: number;
  implied_share_price: number;
  current_price: number;
  upside_pct: number;
}

File: backend/app/models/stock.py
"""Pydantic response models for stock data endpoints."""

from typing import Any

from pydantic import BaseModel


class SearchResult(BaseModel):
    """A single ticker search result."""

    symbol: str
    name: str
    exchange: str | None = None
    type: str | None = None


class SearchResponse(BaseModel):
    """Response for the ticker search endpoint."""

    results: list[SearchResult]


class CompanyProfile(BaseModel):
    """Core company information and metadata."""

    symbol: str
    name: str
    sector: str | None = None
    industry: str | None = None
    country: str | None = None
    website: str | None = None
    description: str | None = None
    market_cap: float | None = None
    employees: int | None = None
    currency: str | None = None
    exchange: str | None = None
    logo_url: str | None = None
    shares_outstanding: float | None = None
    current_price: float | None = None


class PricePoint(BaseModel):
    """A single historical price data point."""

    date: str
    open: float | None = None
    high: float | None = None
    low: float | None = None
    close: float | None = None
    volume: int | None = None


class HistoricalPrices(BaseModel):
    """Response for historical price data."""

    symbol: str
    period: str
    prices: list[PricePoint]


class FinancialStatement(BaseModel):
    """A single period's financial statement as key-value pairs.

    Keys are line-item names (e.g. 'Total Revenue'), values are amounts.
    """

    date: str
    data: dict[str, float | None]


class FinancialStatementResponse(BaseModel):
    """Response for income statement, balance sheet, or cash flow endpoints."""

    symbol: str
    statement_type: str
    statements: list[FinancialStatement]


class KeyRatios(BaseModel):
    """Key financial ratios and valuation metrics for a company."""

    symbol: str
    pe_ratio: float | None = None
    forward_pe: float | None = None
    peg_ratio: float | None = None
    price_to_book: float | None = None
    price_to_sales: float | None = None
    ev_to_ebitda: float | None = None
    profit_margin: float | None = None
    operating_margin: float | None = None
    return_on_equity: float | None = None
    return_on_assets: float | None = None
    debt_to_equity: float | None = None
    current_ratio: float | None = None
    quick_ratio: float | None = None
    dividend_yield: float | None = None
    beta: float | None = None
    fifty_two_week_high: float | None = None
    fifty_two_week_low: float | None = None


class IndustryAverages(BaseModel):
    """Averaged financial metrics across industry peers."""

    industry: str
    peer_count: int
    operating_margin: float | None = None
    tax_rate: float | None = None
    capex_pct_revenue: float | None = None
    revenue_growth: float | None = None


class Projection(BaseModel):
    """A saved projection/scenario from the database."""

    id: str
    ticker: str
    title: str
    data: dict[str, Any]
    created_at: str
    updated_at: str


class ProjectionCreate(BaseModel):
    """Request body to create a new projection."""

    ticker: str
    title: str
    data: dict[str, Any]


class ProjectionUpdate(BaseModel):
    """Request body to update an existing projection."""

    title: str | None = None
    data: dict[str, Any] | None = None
